### DISCUSSION TITLE
Indexing Stalled

### DISCUSSION THREAD

**Discussion by scarolan** [photoprism/photoprism#3646] [Q&A] [answered]
Maybe someone here can help.  I have a large number of photos stored on a NAS drive, which is mounted as the originals folder.  Indexing was zipping along at about 6000 photos per day but it eventually slowed and stopped adding new photos.  When I click the index button it says 'already indexing'.

The container has 2GB of RAM, 4GB of swap, and plenty of extra disk space.  The CPU is not overtaxed and this is all I see in the logs:

```
2023-08-17 06:36:51 INFO index: updated 0 files [2.121178403s]
2023-08-17 06:36:51 ERRO index: already running
2023-08-17 06:36:49 INFO indexing originals...
2023-08-17 06:29:08 WARN purge: already running (start)
```

Have I hit some kind of max limit?  Maybe Photoprism can't handle more than 12,000 files?

**Comment by scarolan** [comment]
So fiddling around and clicking the Cancel button a couple times seems to get it unstuck.  It seems to get some momentum, crawl a thousand or so more photos and then get stuck again.  Oddly the Cancel button never stops the indexing, but it does get it moving again.  Also my 'Restart' button is grayed out so I can't restart the server from the UI.

**Reply by lastzero** [reply]
PhotoPrism can handle millions of files and I've never heard of anything like that. Maybe a problem with the network (file system) since the files you are indexing are not stored locally? You could try NFS instead of CIFS (Windows) or vice versa to see if this helps.

**Reply by lastzero** [reply]
Also, make sure you use MariaDB instead of SQLite and never use network storage for the database files, as this will slow down operations significantly and could even break your index:
- https://docs.photoprism.app/getting-started/faq/#should-i-use-sqlite-mariadb-or-mysql

**Reply by scarolan** [reply]
Thanks @lastzero - it's an older NAS so that may be the bottleneck here.  It is already NFS mounted.  

Is there a safe procedure for restarting Photoprism documented anywhere?  This button remains grayed out.  Should I simply run a `shutdown -r` command from the CLI?

<img width="160" alt="image" src="https://github.com/photoprism/photoprism/assets/403332/61a2f69a-934a-4758-a011-5138ebb54825">

**Reply by lastzero** [reply]
You can simply send a SIGINT signal to the process/container or kill it if that doesn't work. Only the database should be shut down properly if possible, so you don't lose data that hasn't been written to disk yet.

**Reply by lastzero** [reply]
The button works too: It is only active when changes have been made. Simply click twice on one of the check boxes, for example.

**Reply by scarolan** [reply]
Thanks, it looks like there's a systemd script on my container (I'm using LXC on Proxmox), so I did:

```
systemctl photoprism stop
systemctl photoprism start
```

and then kicked off a new indexing run.  From the logs it appears to be MP4 files that are causing the bottleneck, but at least I have a process to get it unstuck now.  Thanks for your help, great app, love the UI and features!

**Reply by lastzero** [reply]
We recently learned that some proxmox users are using a custom version built from source. In this case, the VM most likely also contains different dependencies than our Docker image. As a result, it may behave differently - especially when it comes to video transcoding or RAW/HEIC image conversion.

**Reply by scarolan** [reply, marked as answer]
Good info to have, thanks.  The source-built Proxmox version is definitely an easy way to get started, a single line of bash gets you a Photoprism instance running in an LXC container:

https://github.com/tteck/Proxmox/raw/main/ct/photoprism.sh

I'm pretty sure in my case the bottleneck is the NFS storage.  Running ffmpeg across the wire on large video files seems to take 1-5 minutes depending on the video size.  Thank you for helping me track down the root cause, I'm fine with letting the indexer crawl away and do it's thing as long as it's not completely stuck.

If any Proxmox users come across this thread in the future here are some recommendations:
* Increase your container size to at least 30GB so you have room for metadata and thumbnails, etc
* 512MB of swap is not enough, bump it to 4GB for better performance
* You can run the following command to watch the logs for progress:

```
tail -f /var/lib/photoprism/storage/photoprism.log
```
